crates/kubelet/src/device_plugin_manager/mod.rs

```diff
@@ -0,0 +1,954 @@
+            });
+            Ok(Response::new(Box::pin(
+                tokio_stream::wrappers::ReceiverStream::new(kubelet_update_receiver),
+            )))
+        }
+
+        async fn get_preferred_allocation(
+            &self,
+            _request: Request<PreferredAllocationRequest>,
+        ) -> Result<Response<PreferredAllocationResponse>, Status> {
+            unimplemented!();
+        }
+
+        async fn allocate(
+            &self,
+            _request: Request<AllocateRequest>,
+        ) -> Result<Response<AllocateResponse>, Status> {
+            unimplemented!();
+        }
+
+        async fn pre_start_container(
+            &self,
+            _request: Request<PreStartContainerRequest>,
+        ) -> Result<Response<PreStartContainerResponse>, Status> {
+            Ok(Response::new(PreStartContainerResponse {}))
+        }
+    }
+
+    /// Serves the mock DP
+    async fn run_mock_device_plugin(
+        socket_path: impl AsRef<Path>,
+        devices_receiver: watch::Receiver<Vec<Device>>,
+    ) -> anyhow::Result<()> {
+        let device_plugin = MockDevicePlugin { devices_receiver };
+        let socket = grpc_sock::server::Socket::new(&socket_path).expect("couldnt make dp socket");
+        println!("after creating DP socket");
+        let serv = Server::builder()
+            .add_service(DevicePluginServer::new(device_plugin))
+            .serve_with_incoming(socket);
+        #[cfg(target_family = "windows")]
+        let serv = serv.compat();
+        serv.await.expect("Unable to serve mock device plugin");
+        Ok(())
+    }
+
+    /// Registers the mock DP with the DeviceManager's registration service
+    async fn register_mock_device_plugin(
+        kubelet_socket: String,
+        dp_socket: &str,
+        dp_resource_name: &str,
+    ) -> anyhow::Result<()> {
+        let op = DevicePluginOptions {
+            get_preferred_allocation_available: false,
+            pre_start_required: false,
+        };
+        let channel = grpc_sock::client::socket_channel(kubelet_socket).await?;
+        let mut registration_client = registration_client::RegistrationClient::new(channel);
+        let register_request = tonic::Request::new(RegisterRequest {
+            version: API_VERSION.into(),
+            endpoint: dp_socket.to_string(),
+            resource_name: dp_resource_name.to_string(),
+            options: Some(op),
+        });
+        registration_client
+            .register(register_request)
+            .await
+            .unwrap();
+        Ok(())
+    }
+
+    /// Creates a mock kubernetes API service that the NodePatcher calls to when device plugin resources
+    /// need to be updated in the Node status.
+    /// It verifies the request and always returns a fake Node.
+    /// Returns a client that will reference this mock service and the task the service is running on.
+    /// TODO: Decide whether to test node status
+    pub async fn create_mock_kube_service(
```

Comment:
Shouldn't need to be `pub` in a test

Reply:
This test helper is used for both the DeviceManager and NodePatcher tests. Do you recommend a certain way to share test helpers? Do you prefer tests at the end of each module file or one test section per module?

Reply:
Now that I have thought about it more, this should be fine as it won't be exported anyway unless it is a test. I honestly don't know the best way to share test helpers though

Reply:
I created one public `test_utils` module at the root of the crate.